### PR TITLE
8322881: java/nio/file/Files/CopyMoveVariations.java fails with AccessDeniedException due to permissions of files in /tmp

### PR DESCRIPTION
**Pull request by yaqsun**
testcase: java/nio/file/Files/CopyMoveVariations.java

Method createTempFile() creates "/tmp/file*" that it causes file copy permission issue when running through jtreg.

The method call `Files.move(source, target, options)` by Regular User causes AccessDeniedException due to ` /tmp/file*` is created when running through jtreg.

Create files for method "createTempFile()" passing in the current path(jtreg_test_jdk_java_nio_file_Files_CopyMoveVariations_java/scratch or JTwork/scratch).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8322881](https://bugs.openjdk.org/browse/JDK-8322881): java/nio/file/Files/CopyMoveVariations.java fails with AccessDeniedException due to permissions of files in /tmp (**Bug** - P4)

### Reviewers
 * [Brian Burkhalter](https://openjdk.org/census#bpb) (@bplb - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17235/head:pull/17235` \
`$ git checkout pull/17235`

Update a local copy of the PR: \
`$ git checkout pull/17235` \
`$ git pull https://git.openjdk.org/jdk.git pull/17235/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17235`

View PR using the GUI difftool: \
`$ git pr show -t 17235`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17235.diff">https://git.openjdk.org/jdk/pull/17235.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17235#issuecomment-1874816260)

### CODE REVIEW

**Comment by bplb** [comment]
> Method createTempFile() creates "/tmp/file*" directory that it causes file copy permission issue when running through jtreg.

The method call `Files.createTempFile("file", "dat")` creates a file named "fileXdat" in the default temporary file directory "/tmp" where `X` denotes a large integer. It does not create a directory per se. Could this problem instead be due to permissions on the default temporary file directory "/tmp" on the test system involved?

**Comment by yaqsun** [comment]
> > Method createTempFile() creates "/tmp/file*" directory that it causes file copy permission issue when running through jtreg.
> 
> Could this problem instead be due to permissions on the default temporary file directory "/tmp" on the test system involved?

Yes,  the default temporary file directory "/tmp" causes permission issues.

**Comment by AlanBateman** [comment]
I changed the title of the issue in JBS to make it clear that this is an issue with the test and file permissions and not a bug in createTempFile.

**Comment by bplb** [comment]
> Yes, the default temporary file directory "/tmp" causes permission issues.

Is this issue then perhaps a problem with the system which should not require a change to the test?

**Comment by bplb** [comment]
@yaqsun Sorry, but I corrected a typo in the issue title so you will again need to change the title of this request to match.

**Comment by yaqsun** [comment]
> > Yes, the default temporary file directory "/tmp" causes permission issues.
> 
> Is this issue then perhaps a problem with the system which should not require a change to the test?

The issue is unrelated to system, as the issue is repeated when running with Jtreg command.
```
jtreg/bin/jtreg -dir:test/jdk -testjdk:./** java/nio/file/Files/CopyMoveVariations.java
```

**Comment by yaqsun** [comment]
> @yaqsun Sorry, but I corrected a typo in the issue title so you will again need to change the title of this request to match.

The title have been updated.

**Comment by jaikiran** [comment]
Hello @yaqsun, I just did a search for "Files.createTempFile" under `/test/jdk` directory:

```
grep "Files.createTempFile" ./test/jdk -R
```

There are numerous tests within the JDK which do that call without passing the call a directory under which to create these temporary files, just like this `CopyMoveVariations`. Are those tests passing for you on this system and only this specific one fails?

**Comment by jaikiran** [comment]
Reading through the stacktrace in the linked JBS issue, it appears that it's not the "Files.createTempFile" which is failing but the `Files.move` appears to be failing, unable to move a temporary file (which has `000` posix permission) that was created by the test in the default filesystem's temporary directory, to a non-existent target file on the same filesystem. The copy operation in the `UnixFileSystem` is where this is failing to open the source file. You might have to debug that code a bit to see how that `/tmp` filesystem is setup and different from other systems where this is passing.

**Comment by AlanBateman** [comment]
I suspect the yaqsun runs into with this test because this test is testing file permissions. Most of the other tests that end up using the system tmp directory won't be as sensitive to the file permissions.

**Comment by yaqsun** [comment]
I search testcases about grep "Files.createTempFile" ./test/jdk -R, and test.
Only java/nio/file/Files/CopyMoveVariations.java fail, so the method call `Files.move(source, target, options)` by Regular User causes AccessDeniedException due to ` /tmp/file*` is created when running through jtreg.

```
$ jtreg/bin/jtreg -dir:test/jdk -nativepath:./native-jdk/jdk/jtreg/native -verbose:fail,error,summary -testjdk:./x64-release-jdk jdk/jfr/threading/TestManyVirtualThreads.java jdk/jfr/api/consumer/filestream/TestOrdered.java jdk/jfr/api/consumer/filestream/TestReuse.java jdk/nio/zipfs/ZeroDate.java jdk/nio/zipfs/UpdateEntryTest.java tools/jlink/plugins/GenerateJLIClassesPluginTest.java tools/jar/modularJar/JarToolModuleDescriptorReproducibilityTest.java tools/jimage/JImageListTest.java tools/jimage/JImageExtractTest.java tools/jimage/JImageVerifyTest.java com/sun/management/HotSpotDiagnosticMXBean/DumpThreads.java java/util/zip/ZipFile/ZeroDate.java java/util/concurrent/StructuredTaskScope/StructuredThreadDumpTest.java java/util/Properties/PropertiesStoreTest.java java/util/Properties/StoreReproducibilityTest.java java/io/FileInputStream/NegativeAvailable.java java/io/File/LastModifiedTest.java java/io/File/GetXSpace.java java/io/File/TempDirDoesNotExist.java java/io/PrintStream/FailingConstructors.java java/io/ByteArrayInputStream/ChunkedTransferTo.java java/nio/file/WatchService/UpdateInterference.java java/nio/file/etc/MacVolumesTest.java java/nio/file/Files/DeleteOnClose.java java/nio/file/Files/CopyToNonDefaultFS.java java/nio/file/Files/TemporaryFiles.java java/nio/file/Files/StreamLinesTest.java java/nio/file/Files/CopyInterference.java java/nio/file/Files/StreamTest.java java/nio/file/Files/CallWithInterruptSet.java java/nio/file/Files/BytesAndLines.java java/nio/file/Files/SubstDrive.java java/nio/file/Files/CopyMoveVariations.java java/nio/file/Files/probeContentType/Basic.java java/nio/file/Files/probeContentType/ParallelProbes.java java/nio/file/attribute/PosixFileAttributeView/Basic.java java/nio/file/Path/MacToRealPathWithSM.java java/nio/MappedByteBuffer/Force.java java/nio/channels/FileChannel/MapWithSecurityManager.java java/nio/channels/FileChannel/directio/DirectIOTest.java java/nio/channels/FileChannel/LargeGatheringWrite.java java/nio/channels/FileChannel/Transfer2GPlus.java java/nio/channels/FileChannel/CloseDuringTransfer.java java/nio/channels/FileChannel/TempDirectBuffersReclamation.java java/nio/channels/FileChannel/Position.java java/nio/channels/FileChannel/TransferToAppending.java java/nio/channels/FileChannel/TransferFromExtend.java java/nio/channels/FileChannel/Transfer4GBFile.java java/nio/channels/FileChannel/LoopingTruncate.java java/nio/channels/FileLock/Overlaps.java java/nio/channels/Channels/TransferTo.java java/lang/String/NoReplTest.java java/lang/module/ModuleFinderTest.java java/lang/management/BufferPoolMXBean/Basic.java java/lang/StackTraceElement/SerialTest.java java/net/httpclient/security/Driver.java java/net/httpclient/SmokeTest.java javax/sound/sampled/spi/AudioFileReader/ShortHeader.java javax/sound/sampled/spi/AudioFileReader/AudioFileClose.java
Directory "JTwork" not found: creating
Directory "JTreport" not found: creating
Passed: com/sun/management/HotSpotDiagnosticMXBean/DumpThreads.java
Passed: java/io/ByteArrayInputStream/ChunkedTransferTo.java
Passed: java/io/File/GetXSpace.java
Passed: java/io/File/LastModifiedTest.java
Passed: java/io/File/TempDirDoesNotExist.java
Passed: java/io/FileInputStream/NegativeAvailable.java
Passed: java/io/PrintStream/FailingConstructors.java
Passed: java/lang/management/BufferPoolMXBean/Basic.java
Passed: java/lang/module/ModuleFinderTest.java
Passed: java/lang/StackTraceElement/SerialTest.java
Passed: java/lang/String/NoReplTest.java
Passed: java/net/httpclient/security/Driver.java
Passed: java/net/httpclient/SmokeTest.java
Passed: java/nio/channels/Channels/TransferTo.java
Passed: java/nio/channels/FileChannel/CloseDuringTransfer.java
Passed: java/nio/channels/FileChannel/directio/DirectIOTest.java
Passed: java/nio/channels/FileChannel/LargeGatheringWrite.java
Passed: java/nio/channels/FileChannel/LoopingTruncate.java
Passed: java/nio/channels/FileChannel/MapWithSecurityManager.java
Passed: java/nio/channels/FileChannel/Position.java
Passed: java/nio/channels/FileChannel/TempDirectBuffersReclamation.java
Passed: java/nio/channels/FileChannel/Transfer2GPlus.java
Passed: java/nio/channels/FileChannel/Transfer4GBFile.java
Passed: java/nio/channels/FileChannel/TransferFromExtend.java
Passed: java/nio/channels/FileChannel/TransferToAppending.java
Passed: java/nio/channels/FileLock/Overlaps.java
Passed: java/nio/file/attribute/PosixFileAttributeView/Basic.java
Passed: java/nio/file/Files/BytesAndLines.java
Passed: java/nio/file/Files/CallWithInterruptSet.java
Passed: java/nio/file/Files/CopyInterference.java
--------------------------------------------------
Failed: java/nio/file/Files/CopyMoveVariations.java
Passed: java/nio/file/Files/CopyToNonDefaultFS.java
Passed: java/nio/file/Files/DeleteOnClose.java
Passed: java/nio/file/Files/probeContentType/Basic.java
Passed: java/nio/file/Files/probeContentType/ParallelProbes.java
Passed: java/nio/file/Files/StreamLinesTest.java
Passed: java/nio/file/Files/StreamTest.java
Passed: java/nio/file/Files/TemporaryFiles.java
Passed: java/nio/file/WatchService/UpdateInterference.java
Passed: java/nio/MappedByteBuffer/Force.java
Passed: java/util/concurrent/StructuredTaskScope/StructuredThreadDumpTest.java
Passed: java/util/Properties/PropertiesStoreTest.java
Passed: java/util/Properties/StoreReproducibilityTest.java
Passed: java/util/zip/ZipFile/ZeroDate.java
Passed: javax/sound/sampled/spi/AudioFileReader/AudioFileClose.java
Passed: javax/sound/sampled/spi/AudioFileReader/ShortHeader.java
Passed: jdk/jfr/api/consumer/filestream/TestOrdered.java
Passed: jdk/jfr/api/consumer/filestream/TestReuse.java
Passed: jdk/jfr/threading/TestManyVirtualThreads.java
Passed: jdk/nio/zipfs/UpdateEntryTest.java
Passed: jdk/nio/zipfs/ZeroDate.java
Passed: tools/jar/modularJar/JarToolModuleDescriptorReproducibilityTest.java
Passed: tools/jimage/JImageExtractTest.java
Passed: tools/jimage/JImageListTest.java
Passed: tools/jimage/JImageVerifyTest.java
Passed: tools/jlink/plugins/GenerateJLIClassesPluginTest.java
Test results: passed: 55; failed: 1
```

**Comment by sendaoYan** [comment]
Another solution: https://github.com/openjdk/jdk/pull/17606

**Comment by yaqsun** [comment]
Please review my pr.
Thank you very much.

**Comment by AlanBateman** [comment]
> Please review my pr. 

I don't think the exchange so far has clearly established the issue. It is likely related to the permissions on /tmp on this system but maybe there is a umask setting in the picture too. In any case, I think it's better if this test uses the work directory rather than /tmp.  You'll need to change the static initializer where it first tests if underlying file store support POSIX file permissions. Also in the op method it would be nicer if you used something like `Path dir = Path.of(".");` so you have a Path to the directory for the 4 places that use it.

**Comment by bplb** [comment]
I have not been able to reproduce the failure. I am of the opinion however that the problem is due to something local to the test system, e.g., permissions on `/tmp`, a `umask` setting, etc. That said, I think it's good to change the test directory from `/tmp` to `.` so the change in 034324c0df286203b3b718898c9538f565fb6232 is fine. I will approve this PR tomorrow unless there are objections.

File: test/jdk/java/nio/file/Files/CopyMoveVariations.java [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
```
Comment (bplb) [inline test/jdk/java/nio/file/Files/CopyMoveVariations.java:2]: The copyright needs to be `2023, 2024`, not just `2024`.

File: test/jdk/java/nio/file/Files/CopyMoveVariations.java [hunk 2]
```diff
@@ -72,7 +72,7 @@ enum PathType {
     static {
         Path tmp = null;
         try {
-            tmp = Files.createTempFile("this", "that");
+            tmp = Files.createTempFile(Path.of("."), "this", "that");
```
Comment (bplb) [inline test/jdk/java/nio/file/Files/CopyMoveVariations.java:75]: Could you change `tmp` to be a constant, for example
```
    private static final Path CURRENT_DIR;
```
(or `CWD`) and change `current` to be `CURRENT_DIR`?

Comment (yaqsun) [reply test/jdk/java/nio/file/Files/CopyMoveVariations.java]: > Could you change `tmp` to be a constant, for example
> 
> ```
>     private static final Path CURRENT_DIR;
> ```

```
74         try {
75             currentDir = Files.createTempFile(Path.of("."), "this", "that");
76             SUPPORTS_POSIX_PERMISSIONS =
77                 Files.getFileStore(currentDir).supportsFileAttributeView("posix");
78         } catch (IOException cause) {
79             throw new UncheckedIOException(cause);
80         } finally {
81             if (currentDir != null) {
```

After change tmp to be a constant currentDir,  variable currentDir can not be initialized in the finally block.

```
private static final Path currentDir;
```

**Comment by yaqsun** [comment]
I ran an experiment on existing machines.
Running on the following environment can be replicated.

```
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"
```

and 

```
NAME="CentOS Linux"
VERSION="8"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="8"
PLATFORM_ID="platform:el8"
PRETTY_NAME="CentOS Linux 8"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:8"
HOME_URL="https://centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"
CENTOS_MANTISBT_PROJECT="CentOS-8"
CENTOS_MANTISBT_PROJECT_VERSION="8"
```

But running on the following environment can not be replicated.

```
NAME="Ubuntu"
VERSION="18.04.6 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04.6 LTS"
VERSION_ID="18.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=bionic
UBUNTU_CODENAME=bionic
```

@AlanBateman @bplb Thank you for reminding me.

**Pull request closed without merging**